From b3b48836590306ecc9a8c18c852b8e134f8ef9d9 Mon Sep 17 00:00:00 2001
From: Shadi Ammouri <shadi@marvell.com>
Date: Tue, 16 Aug 2016 17:38:18 +0300
Subject: [PATCH 0425/1345] syscon: a80x0: a70x0: Add support for hclk clock
 source

commit  50f29cc18f3330cf1f8edc06c3bb4abc99008716 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

This clock is used by AXI-monitor for profiling counters.

Change-Id: Ib77907bd83b7d75c6c443c4dbe7580f5fb2b344f
Signed-off-by: Shadi Ammouri <shadi@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31987
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../arm/marvell/ap806-system-controller.txt        |    5 +-
 arch/arm64/boot/dts/marvell/armada-ap806.dtsi      |    2 +-
 drivers/clk/mvebu/ap806-system-controller.c        |   64 ++++++++++++++++++--
 3 files changed, 62 insertions(+), 9 deletions(-)

diff --git a/Documentation/devicetree/bindings/arm/marvell/ap806-system-controller.txt b/Documentation/devicetree/bindings/arm/marvell/ap806-system-controller.txt
index 8968371..702a7be 100644
--- a/Documentation/devicetree/bindings/arm/marvell/ap806-system-controller.txt
+++ b/Documentation/devicetree/bindings/arm/marvell/ap806-system-controller.txt
@@ -14,6 +14,7 @@ a number of clocks:
  - 1: clock of CPU cluster 1
  - 2: fixed PLL at 1200 Mhz
  - 3: MSS clock, derived from the fixed PLL
+ - 4: AP HCLK, derived from sample-at-reset configuration.
 
 Required properties:
 
@@ -22,7 +23,7 @@ Required properties:
  - reg: register area of the AP806 system controller
  - #clock-cells: must be set to 1
  - clock-output-names: must be defined to:
-    "ap-cpu-cluster-0", "ap-cpu-cluster-1", "ap-fixed", "ap-mss"
+    "ap-cpu-cluster-0", "ap-cpu-cluster-1", "ap-fixed", "ap-mss", "ap-hclk"
 
 Example:
 
@@ -30,6 +31,6 @@ Example:
 		compatible = "marvell,ap806-system-controller", "syscon";
 		#clock-cells = <1>;
 		clock-output-names = "ap-cpu-cluster-0", "ap-cpu-cluster-1",
-				     "ap-fixed", "ap-mss";
+				     "ap-fixed", "ap-mss", "ap-hclk";
 		reg = <0x6f4000 0x1000>;
 	};
diff --git a/arch/arm64/boot/dts/marvell/armada-ap806.dtsi b/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
index 2bc7c2b..4ec553b 100644
--- a/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
@@ -358,7 +358,7 @@
 				clock-output-names = "ap-cpu-cluster-0",
 						     "ap-cpu-cluster-1",
 						     "ap-fixed", "ap-mss",
-						     "ap-emmc";
+						     "ap-emmc", "ap-hclk";
 				reg = <0x6f4000 0x1000>;
 			};
 		};
diff --git a/drivers/clk/mvebu/ap806-system-controller.c b/drivers/clk/mvebu/ap806-system-controller.c
index 9e90be7..eb76005 100644
--- a/drivers/clk/mvebu/ap806-system-controller.c
+++ b/drivers/clk/mvebu/ap806-system-controller.c
@@ -23,7 +23,7 @@
 #define AP806_SAR_REG			0x400
 #define AP806_SAR_CLKFREQ_MODE_MASK	0x1f
 
-#define AP806_CLK_NUM			5
+#define AP806_CLK_NUM			6
 
 static struct clk *ap806_clks[AP806_CLK_NUM];
 
@@ -34,7 +34,7 @@
 
 static int ap806_syscon_clk_probe(struct platform_device *pdev)
 {
-	unsigned int freq_mode, cpuclk_freq;
+	unsigned int freq_mode, cpuclk_freq, hclk_freq;
 	const char *name, *fixedclk_name;
 	struct device_node *np = pdev->dev.of_node;
 	struct regmap *regmap;
@@ -97,14 +97,52 @@ static int ap806_syscon_clk_probe(struct platform_device *pdev)
 		dev_err(&pdev->dev, "invalid SAR value\n");
 	}
 
+	/* Get HCLK frequency */
+	switch (freq_mode) {
+	case 0x4:
+	case 0x10:
+	case 0x14:
+	case 0x19 ... 0x1D:
+		hclk_freq = 400;
+		break;
+	case 0xC:
+		hclk_freq = 600;
+		break;
+	case 0xD:
+	case 0x16:
+		hclk_freq = 525;
+		break;
+	case 0xB:
+	case 0xE:
+	case 0xF:
+		hclk_freq = 450;
+		break;
+	case 0x12:
+	case 0x13:
+	case 0x17:
+		hclk_freq = 325;
+		break;
+	case 0x11:
+	case 0x15:
+		hclk_freq = 800;
+		break;
+	case 0x18:
+		hclk_freq = 650;
+		break;
+	default:
+		hclk_freq = 0;
+		pr_err("invalid SAR value\n");
+	}
+
 	/* Convert to hertz */
 	cpuclk_freq *= 1000 * 1000;
+	hclk_freq *= 1000 * 1000;
 
 	/* CPU clocks depend on the Sample At Reset configuration */
 	of_property_read_string_index(np, "clock-output-names",
 				      0, &name);
 	ap806_clks[0] = clk_register_fixed_rate(&pdev->dev, name, NULL,
-						CLK_IS_ROOT, cpuclk_freq);
+						0, cpuclk_freq);
 	if (IS_ERR(ap806_clks[0])) {
 		ret = PTR_ERR(ap806_clks[0]);
 		goto fail0;
@@ -112,7 +150,7 @@ static int ap806_syscon_clk_probe(struct platform_device *pdev)
 
 	of_property_read_string_index(np, "clock-output-names",
 				      1, &name);
-	ap806_clks[1] = clk_register_fixed_rate(&pdev->dev, name, NULL, CLK_IS_ROOT,
+	ap806_clks[1] = clk_register_fixed_rate(&pdev->dev, name, NULL, 0,
 						cpuclk_freq);
 	if (IS_ERR(ap806_clks[1])) {
 		ret = PTR_ERR(ap806_clks[1]);
@@ -122,7 +160,7 @@ static int ap806_syscon_clk_probe(struct platform_device *pdev)
 	/* Fixed clock is always 1200 Mhz */
 	of_property_read_string_index(np, "clock-output-names",
 				      2, &fixedclk_name);
-	ap806_clks[2] = clk_register_fixed_rate(&pdev->dev, fixedclk_name, NULL, CLK_IS_ROOT,
+	ap806_clks[2] = clk_register_fixed_rate(&pdev->dev, fixedclk_name, NULL,
 						0, 1200 * 1000 * 1000);
 	if (IS_ERR(ap806_clks[2])) {
 		ret = PTR_ERR(ap806_clks[2]);
@@ -155,7 +193,21 @@ static int ap806_syscon_clk_probe(struct platform_device *pdev)
 		}
 	}
 
-	of_clk_add_provider(np, of_clk_src_onecell_get, &ap806_clk_data);
+	of_property_read_string_index(np, "clock-output-names",
+				      5, &name);
+	ap806_clks[5] = clk_register_fixed_rate(NULL, name, NULL, 0,
+						dclk_freq);
+	of_property_read_string_index(np, "clock-output-names",
+				      6, &ringclk_name);
+	ap806_clks[6] = clk_register_fixed_rate(NULL, ringclk_name, NULL, 0,
+						ringclk_freq);
+
+	/* External ring is 1/2 of ring clk */
+	of_property_read_string_index(np, "clock-output-names",
+				      7, &name);
+	ap806_clks[7] = clk_register_fixed_factor(NULL, name, ringclk_name,
+						  0, 1, 2);
+
 	ret = of_clk_add_provider(np, of_clk_src_onecell_get, &ap806_clk_data);
 	if (ret)
 		goto fail_clk_add;
-- 
1.7.9.5

